### PR TITLE
dask: `Data.array` and friends

### PR DESCRIPTION
**Pull request by davidhassell**

### CODE REVIEW

**Review by sadielbartholomew — APPROVED** [review]
A few very minor comments, but otherwise I am really happy with this. Please merge when you think it is ready.

File: cf/data/utils.py [hunk 1]
```diff
@@ -61,177 +67,258 @@ def _is_numeric_dtype(array):
...
+            one.
+
+    **Examples**
+
+    >>> import dask.array as da
+    >>> d = da.arange(8).reshape(2, 4)
+    >>> print(d.compute())
+    [[0 1 2 3]
+     [4 5 6 7]]
+    >>> cf.data.utils.first_non_missing_value(d)
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/data/utils.py:253]: Great idea. Will do.

[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/data/utils.py:253]: Done 6a5118990

File: cf/data/utils.py [hunk 2]
```diff
@@ -61,177 +67,258 @@ def _is_numeric_dtype(array):
...
+            value of the array, thereby removing the need to find it
+            by inspection of *a*, which may be expensive. By default
+            the first non-missing value is found from *a*.
 
-            dask array, `Units`
-                A new dask array containing reference times, and its
-                units.
+    :Returns:
+
+        `dask.array.Array`, `Units`
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/data/utils.py]: Is https://github.com/NCAS-CMS/cf-python/pull/353/commits/ac52fe6849d2c5be0564dd1259a0afea26c63ea4 OK? Happy to use some words if you prefer!

File: cf/data/utils.py [hunk 3]
```diff
@@ -61,177 +67,258 @@ def _is_numeric_dtype(array):
...
+
+        method: `str`, optional
+            Select the method used to find the first non-missing
+            value.
 
-    array: dask array
-        The array to be inspected.
+            The default ``'index'`` method evaulates sequentially
+            elements of the flattened array. The ``'mask'`` method
+            finds the first non-missing value from the mask of the
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/data/utils.py]: Indeed!. How about https://github.com/NCAS-CMS/cf-python/pull/353/commits/38e8c452d53aa3977905e745e904258d3b3d7dc3?

File: cf/data/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply cf/data/utils.py]: Yes that's fine too (and simpler)! Thanks, marking as resolved.

File: cf/data/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply cf/data/utils.py]: Perfect, thanks.

File: cf/data/data.py [hunk 4]
```diff
@@ -2944,6 +3001,51 @@ def ceil(self, inplace=False, i=False):
...
+
+        `array` causes all delayed operations to be computed.
+
+        .. versionadded:: TODODASK
+
+        .. seealso:: `persist`, `array`, `datetime_array`
+
+        :Returns:
+
+            `numpy.ndarray`
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/data/data.py:3027]: Masked arrays are subclasses of `np.ndarray`, so I think what we have catches all. This applies to everywhere we use this pattern.

```python
>>> isinstance(np.ma.array(9), np.ndarray)
True
```

[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply cf/data/data.py:3027]: Thanks for clarifying, and based on your comments I agree that it's fine as-is, and I have noted the association for future PRs and reviewing. I think that's everything addressed in terms of feedback now: all good to merge!

**Pull request merged**